src/upchunk.ts

```diff
@@ -314,118 +494,154 @@ export class UpChunk  {
+
+        // Re-estimate the total number of chunks, by adding the completed
+        // chunks to the remaining chunks
+        const remainingChunks =
+          (this.file.size - this.nextChunkRangeStart) / this.chunkByteSize;
+        this.totalChunks = Math.ceil(this.chunkCount + remainingChunks);
+      }
+
+      return true;
+    };
+
+    // What to do if a chunk upload failed, potentially after retries
+    const failedChunkUploadCb = async (res: XhrResponse, _chunk?: Blob) => {
+      // Side effects
+      this.dispatch('error', {
+        message: `Server responded with ${
+          (res as XhrResponse).statusCode
+        }. Stopping upload.`,
+        chunk: this.chunkCount,
+        attempts: this.attemptCount,
+      });
+
+      return false;
+    };
+
+    // What to do if a chunk upload failed but is retriable and hasn't exceeded retry
+    // count
+    const retriableChunkUploadCb = async (
+      _res: XhrResponse | undefined,
+      _chunk?: Blob
+    ) => {
+      // Side effects
       this.dispatch('attemptFailure', {
         message: `An error occured uploading chunk ${this.chunkCount}. ${
           this.attempts - this.attemptCount
         } retries left.`,
         chunkNumber: this.chunkCount,
         attemptsLeft: this.attempts - this.attemptCount,
       });
-      return;
-    }
 
-    this.dispatch('error', {
-      message: `An error occured uploading chunk ${this.chunkCount}. No more retries, stopping upload`,
-      chunk: this.chunkCount,
-      attempts: this.attemptCount,
-    });
+      return new Promise<boolean>((resolve) => {
+        setTimeout(async () => {
+          // Handle mid-flight _paused/offline cases here by storing the
+          // "still retriable but yet to be uploaded chunk" in state.
+          // See also: `sendChunks()`
+          if (this._paused || this.offline) {
+            this.pendingChunk = chunk;
+            resolve(false);
+            return;
+          }
+          const chunkUploadSuccess = await this.sendChunkWithRetries(chunk);
+          resolve(chunkUploadSuccess);
+        }, this.delayBeforeAttempt * 1000);
+      });
+    };
+
+    let res: XhrResponse | undefined;
+    try {
+      this.attemptCount = this.attemptCount + 1;
+      this.lastChunkStart = new Date();
+      res = await this.sendChunk(chunk);
+    } catch (_err) {
+      console.warn('DEBUG: Caught an error: %j', _err);
+      // this type of error can happen after network disconnection on CORS setup
+    }
+    const options = {
+      retryCodes: this.retryCodes,
+      attemptCount: this.attemptCount,
+      attempts: this.attempts,
+    };
+    // console.log('res', res, 'options', options);
+    if (isSuccessfulChunkUpload(res, options)) {
+      return await successfulChunkUploadCb(res, chunk);
+    }
+    if (isFailedChunkUpload(res, options)) {
+      return await failedChunkUploadCb(res, chunk);
+    }
+    // Retriable case
+    return await retriableChunkUploadCb(res, chunk);
```

Comment:
do we want to await here to just return the promises of the callback functions to be awaited by the consumer of `sendChunkWithRetries`?

Reply:
ah so to rephrase: `await` is likely redundant here? That sounds right. I'll refactor and re-test and remove if 👌. Good call.

Reply:
yeah, having the extra await will incur an extra tick, which may not be necessary. I don't think I noticed any other cases, but might be worth double-checking that we're only awaiting at the "terminal" points and dropping from cases like these.